edge-hub/watchdog/src/child.rs

```diff
@@ -0,0 +1,113 @@
+use std::{
+    process::{Child, Command, Stdio},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+    thread::{sleep, JoinHandle},
+    time::Duration,
+};
+
+use anyhow::{Context, Result};
+use nix::{
+    sys::signal::{self, Signal},
+    unistd::Pid,
+};
+use tracing::{error, info};
+
+const PROCESS_SHUTDOWN_TOLERANCE_SECS: Duration = Duration::from_secs(60);
+const PROCESS_POLL_INTERVAL_SECS: Duration = Duration::from_secs(1);
+
+pub struct ChildProcess {
+    name: String,
+    process: Child,
+}
+
+impl ChildProcess {
+    pub fn new(name: String, process: Child) -> Self {
+        Self { name, process }
+    }
+
+    pub fn is_running(&mut self) -> bool {
+        match self.process.try_wait() {
+            Ok(status) => status.is_none(),
+            Err(e) => {
+                error!("Error while polling {} process status. {}", self.name, e);
+                false
+            }
+        }
+    }
+
+    pub fn shutdown_if_running(&mut self) {
+        if self.is_running() {
+            info!("Terminating {} process", self.name);
+            self.send_signal(Signal::SIGTERM);
+        } else {
+            info!("{} process has stopped", self.name)
+        }
+
+        self.wait_for_exit();
+
+        if self.is_running() {
+            info!("Killing {} process", self.name);
+            self.send_signal(Signal::SIGKILL);
+        }
+
+        self.wait_for_exit();
+    }
+
+    fn send_signal(&mut self, signal: Signal) {
+        info!("Sending {} signal to {} process", signal, self.name);
+
+        /* Casting process.id() from u32 to i32 cannot wrap. The max pid for linux is ~4 million. The max value for i32 is 2,147,483,647.
```

Comment:
nit: I don't think it is a rule, but if you could change the comment line width to something around 80 symbols, @vadim-kovalyov will really appreciate.
Also could you switch to `//` type of comment event for multiline please?